walk/walk.go

```diff
@@ -161,17 +161,37 @@ func visit(c *config.Config, cexts []config.Configurer, knownDirectives map[stri
 	c = configure(cexts, knownDirectives, c, rel, f)
 	wc := getWalkConfig(c)
 
+	// Collect gitignore files in this directory.
+	if _, hasIgnore := trie.children[".gitignore"]; hasIgnore {
+		ignoreFilePath := path.Join(c.RepoRoot, rel, ".gitignore")
+		if ignoreReader, ignoreErr := os.Open(ignoreFilePath); ignoreErr == nil {
+			wc.loadGitIgnore(rel, ignoreReader)
+		}
+	}
+
 	if wc.isExcluded(rel) {
 		return
 	}
 
+	// Append an additional element to the relParts array to allow swapping the trailing entry
+	// in each iteration of the loop below.
+	entParts := append(relParts, ".")
+
 	var subdirs, regularFiles []string
 	for _, ent := range ents {
 		base := ent.Name()
 		entRel := path.Join(rel, base)
 		if wc.isExcluded(entRel) {
 			continue
 		}
+
+		entParts[len(entParts)-1] = base
+
+		// Skip git-ignored when the gitignore option is enabled.
+		if wc.isGitIgnored(entParts, ent.IsDir()) {
```

Comment:
Note this is its own conditional and not embedded within `isExcluded` for a few reasons:
* I don't think .gitignore can exclude the parent directory it is within the way `# gazelle:exclude *` can (that is the reason we must invoke `isExcluded` both above and within this `for` loop).
* `isGitIgnored` requires the path as `[]string` and needs to know `IsDir()`, I don't think we should override `isExcluded` with more params